src/agent/onefuzz-supervisor/src/work.rs

```diff
@@ -171,38 +170,44 @@ impl WorkQueue {
 
         // Now we've had a chance to ensure our SAS URL is fresh. For any other
         // error, including another auth error, bail.
-        let msg = msg.context("unable to check work queue")?;
-
-        let result = match msg {
-            Some(msg) => {
-                let work_set =
-                    serde_json::from_slice(msg.data()).context("unable to parse WorkSet")?;
-                let receipt = Receipt(msg.receipt);
-                Some(Message { receipt, work_set })
-            }
-            None => None,
-        };
-        Ok(result)
-    }
+        let msg = msg.context("failed to pop message")?;
 
-    pub async fn claim(&mut self, receipt: Receipt) -> Result<()> {
-        let receipt = receipt.0;
+        if msg.is_none() {
+            return Ok(None);
+        }
 
-        let result = self.queue.delete(receipt.clone()).await;
+        let queue_message = msg.unwrap();
+        let work_set: WorkSet = queue_message.get()?;
+        let msg = Message {
+            queue_message: Some(queue_message),
+            work_set,
+        };
 
-        // If we had an auth err, renew our registration and retry once, in case
-        // it was just due to a stale SAS URL.
-        if let Err(err) = &result {
-            if is_auth_error(err) {
-                self.renew().await.context("unable to renew registration")?;
-                self.queue
-                    .delete(receipt)
-                    .await
-                    .context("unable to claim work from queue")?;
+        Ok(Some(msg))
+    }
+
+    pub async fn claim(&mut self, message: Message) -> Result<WorkSet> {
```

Comment:
Why have `claim()` return a work set? If the caller needs it, they can do a partial move or `clone()` it.

Reply:
In my mind, the argument to this method should just be the metadata required to claim the message (i.e. finalize the queue pop), and that + the message payload could be in their own struct.

Reply:
when claiming the message we call delete on the inner queue_message which takes ownership of the message. 
So the Message argument is not usable after the call to this. Returning the work_set allows us to avoid a clone. The caller still has the choice of not using the return value.

Reply:
This comment is _not_ a PR blocker, but I just want to clarify my suggestion. Feel free to resolve once read.

Taking ownership of the inner `queue_message` field is not a problem (unless there is something new to the situation I missed!). Until this PR, we already did what I am suggesting (with the previous version of the field, which held a `Receipt`). We apparently had a `work_set.clone()` in one of the branches, but it is actually redundant.

So if we maintained / switch back to a similar API, it would look like:
```rust
pub struct Metadata(Option<QueueMessage>);

pub struct Message {
    pub metadata: Metadata,  // Before, this was always a `Receipt`.
    pub work_set: WorkSet,
}

// Assume the caller splits off `Metadata` from the `Message.metadata` field.
// Now `claim()` only takes ownership of what it actually needs, 
// and doesn't even need to "know" the message payload type.
pub async fn claim(&mut self, metadata: Metadata) -> Result<()>;
```
Then we can handle the `Message`/`claim()` like we did before:
```rust
// Move one field out of the struct.
// This is fine as long as we only try to use other (un-moved) fields later on, 
// and not the entire struct. Note we don't _need_ this temp variable.
let work_set = message.work_set;

// Use the other, un-moved field. We can now access `work_set` as needed, since it is in scope.
match self.work_queue.claim(message.metadata) {
    // Handle the different cases.
    Ok(..) => {
        do_something_with(work_set);
    }
    Err(..) => {
        print_warning_about(work_set);
    }
}
```